File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        generate("", n, 0, 0, result);
        return result;
    }
    
    private void generate(String current, int n, int open, int close, List<String> result) {
        if (open == n && close == n) {
            result.add(current);
            return;
        }
        
        if (open < n) {
            generate(current + "(", n, open+1, close, result);
        }
        
        if (close < open) {
            generate(current + ")", n, open, close+1, result);
        }
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []
        self.generate("", n, 0, 0, result)
        return result
    
    def generate(self, current: str, n: int, open: int, close: int, result: List[str]):
        if open == n and close == n:
            result.append(current)
            return
        
        if open < n:
            self.generate(current + "(", n, open+1, close, result)
        
        if close < open:
            self.generate(current + ")", n, open, close+1, result)
